validation/type_references_constraint.go

```diff
@@ -72,23 +80,22 @@ func (c *TypeReferencesConstraint) checkTypeIsDefined(t, referencedType string,
 	return typeDef, nil
 }
 
-func (c *TypeReferencesConstraint) checkCircularDependency(typeFactory goldi.TypeFactory, typeID string, container *goldi.Container) error {
+func (c *TypeReferencesConstraint) checkCircularDependency(typeFactory goldi.TypeFactory, typeID string, container *goldi.Container, g *Graph) error {
```

Comment:
With these changes in place all the function does is to build the entire dependency graph recursively and there are no more checks going on here. Lets rename this function to `buildDependencyGraph`. With the new name it is much more intuitive why we have the conditional in the middle:

```go
if g.IsNodePresent(referencedTypeID) {
	g.AddEdge(typeID, referencedTypeID)
		continue
}
```